jetty-jaspi/src/test/java/org/eclipse/jetty/security/jaspi/JaspiTest.java

```diff
@@ -86,10 +89,31 @@ protected UserPrincipal loadUserInfo(String username)
         }
     }
 
+    @BeforeAll
+    public static void beforeAll() throws Exception
+    {
+        AuthConfigFactory factory = new BasicAuthConfigFactory();
+        
```

Comment:
We also need to come up with a way of configuring this in a jetty xml file, as this is how jetty modules are enabled. Take a look at the `jetty-jaspi/src/main/config/modules/jaspi.mod` file - it references an xml that will be put on jetty's execution path to configure jaspi. Have a read of https://www.eclipse.org/jetty/documentation/jetty-11/operations-guide/index.html#og-modules to find out about jetty modules.

Reply:
Ok, the previous implementation relied on the old Geronimo implementation (so they'd both need to be set up) and the xml file reflects that. I'm not sure that's appropriate to replicate into the jetty project, presumably (unless geronimo upgrades at some point) people would pick the glassfish one instead (which has a different configuration format (not-xml), or another implementation.  Alternatively a new configuration format would need to be implemented in jetty if one wants to configure it through xml or what not.

Reply:
Perhaps an alternative would be to provide an example of a <Configure> xml file?

Reply:
I was thinking we could provide an xml file for the `BasicAuthConfigFactory`, and parameterize the necessary args (classnames, map values etc).  

To allow users to provide their own `AuthConfigFactory` implementation (reasonable?) maybe it would be a` jaspi.mod` to get the jars onto the classpath and maybe a `jaspi-basic.mod` to get the `BasicAuthConfigFactory` xml config file onto the classpath?

Reply:
this can be a list of ServerAuthModules though, as the JaspiTest shows, it's 'possible' to have different ServerAuthModules for different context paths (and possibly by different logical servers). Though I'd imagine it would be rare - at least in a microservice world.  

If we settle for a single ServerAuthModule supported by default in a module we could parameterize classname of ServerAuthModule impl and context path (possibly also logical server, but I don't see it used so far).  Maybe support realm, but that's jetty specific configuration

Reply:
Maybe the jaspi-basic.xml file would be as simple as creating the BasicAuthConfigFactory, and assigning it an `id` so it could be referenced in user-supplied xml files to configure the various ServerAuthModules. As you said previously, maybe we need to provide an example, commented out? Although, I kinda hate to do that, as nothing rots as quickly as commented out code/xml ;)

Reply:
ok, I'm not familiar with the modules configuration so will need to look into it and give it a shot, probably later tonight

Reply:
I'm out of my depth here, but there are several things that need to be done.  I've tried to illustrate using the following mock jaspi-basic.xml file

`<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_10_0.dtd">

<Configure id="context" class="org.eclipse.jetty.webapp.WebAppContext">

  <!-- Step 1: init AuthConfigFactory for jakarta.   -->
  <New id="jaspiAuthConfigFactory" class="org.eclipse.jetty.jaspi.BasicAuthConfigFactory" />
  <Call class="jakarta.security.auth.message.config.AuthConfigFactory" name="setFactory">
    <Arg><Ref refid="jaspiAuthConfigFactory" /></Arg>
  </Call>
  
  <!-- Step 2: configure AuthConfigProvider with appropriate ServerAuthModules -->
  <Ref refid="jaspiAuthConfigFactory">
	<Call name="registerConfigProvider">
          <Arg type="String">org.eclipse.jetty.security.jaspi.BasicAuthModule</Arg>
          <Arg>
              <Map>
                  <Entry>
                      <Item>AppContextID</Item>
                      <Item>server /other</Item>
                  </Entry>
                  <Entry>
                      <Item>org.eclipse.jetty.security.jaspi.modules.RealmName</Item>
                      <Item>TestRealm</Item>
                  </Entry>
              </Map>
          </Arg>
          <Arg type="String">HttpServlet</Arg>
          <Arg type="String">server /ctx</Arg>
          <Arg type="String">a test provider</Arg>
	</Call>
	<Call name="registerConfigProvider">
            <Arg type="String">org.eclipse.jetty.security.jaspi.HttpHeaderAuthModule</Arg>
            <Arg>
                <Map>
                    <Entry>
                        <Item>AppContextID</Item>
                        <Item>server /other</Item>
                    </Entry>
                </Map>
            </Arg>
            <Arg type="String">HttpServlet</Arg>
            <Arg type="String">server /other</Arg>
            <Arg type="String">another test provider</Arg>
	</Call>
  </Ref>
  
  <!-- Step 3: connect JASPI AuthConfigFactory to WebAppContext -->
  <Set name="securityHandler">
    <New class="org.eclipse.jetty.security.ConstraintSecurityHandler">
      <Set name="loginService">
        <New class="org.eclipse.jetty.plus.jaas.JAASLoginService">
          <Set name="name">JAASRealm</Set>
          <Set name="loginModuleName">jaas</Set>
        </New>
      </Set>
  
      <Set name="authenticatorFactory">
        <New class="org.eclipse.jetty.security.jaspi.JaspiAuthenticatorFactory" />
      </Set>
    </New>
  </Set>
  
</Configure>
`

**Step 1**:  This sets up the BasicAuthConfigFactory to be de default AuthConfigFactory.  Can alternatively be done through system property  "authconfigprovider.factory" defined by jakarta.security.auth.message.config.AuthConfigFactory#DEFAULT_FACTORY_SECURITY_PROPERTY and referenced from jakarta...AuthConfigFactory#getFactory

Should one wish to use a different AuthConfigFactory (e.g. glassfish or another impl), then it might be better to provide this through a system property (with a default value if not set otherwise)

**Step 2**:  This is custom per app, but is usable across different AuthConfigFatory implementations (though most provide another means of configuration as well - such as the old geronimo jaspi script).  I've emulated the test Server AuthModule for good measure, but other portable ServerAuthModules could be configured here depending on need

**Step 3**: Finally, it all needs to be connected back to Jetty WebAppContext.  I don't see webappcontext being a default id in the module system, but "context" seems to have been used in the examples.  Not sure how this is all wired together

Reply:
My question is then, how would this best be separated into resuable/configurable components?

Reply:
OK, so the `jaspi-basic.xml` file needs to configure stuff at the `Server` level, because when it runs, the `WebAppContext` won't exist yet. IIUC the `BasicAuthConfigFactory` can be a shared resource amongst all contexts, right?  If so, then the `jaspi-basic.xml` file would look like:

```xml
<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://www.eclipse.org/jetty/configure_10_0.dtd">

<Configure id="Server" class="org.eclipse.jetty.server.Server">

  <!-- ===================================================================== -->
  <!-- Configure a factory for Jaspi                                         -->
  <!-- ===================================================================== -->
  <Call name="addBean">
   <Arg>
        <New id="jaspiAuthConfigFactory" class="org.eclipse.jetty.jaspi.BasicAuthConfigFactory" />
   </Arg>
  </Call>
  <Call class="jakarta.security.auth.message.config.AuthConfigFactory" name="setFactory">
      <Arg><Ref refid="jaspiAuthConfigFactory" /></Arg>
  </Call>
</Configure>
```

You can remove all of **Step 3**.  The `ConstraintSecurityHandler` is the default that will be used; the `SecurityHandler` will find whatever `LoginService` has been configured (or the user can define their own in their own context xml file - that's a documented part of jetty). As for the `JaspiAuthenticatorFactory`, you don't need to configure it for each webapp. Instead, you put a `META-INF/services/` definition file in `jetty-jaspi/src/main/resources` for the `JaspiAuthenticatorFactory` and jetty will find it and hook it up to the `SecurityHandler` when it starts. 

So, AFAICT it is just **Step 2** that needs to be done per `WebAppContext` in a user-defined context xml file. So in that case, we could do the same as the `openid.mod` does, and copy a commented-out (urgh) example xml file into the `$jetty.base/etc` directory that the user can modify, see: https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-openid/src/main/config/modules/openid.mod#L15  and

Reply:
Ok, sounds good. 
->Step 1: this means that if anyone wants to use a different provider (atm only glassfish has one compatible), they'd have to create a new module definition to match this part (replacing BasicAuthConfigFactory).  Perhaps the classname should be parameterized somehow?
->Step 2: added example code, ( note this must always be tweaked to match chosen ServerAuthModules,as the two implementations given are only availlable in unittest scope)